File: challenge/leet.gcc/easy/q268/q268.cpp
/*
 * 268. Missing Number
 * https://leetcode.com/problems/missing-number/
 */

#include <vector>
#include <iostream>
#include <numeric>
#include <string>

using namespace std;

class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int sz = nums.size();
        int p = 0, sk = 0, bk = sz;

        while (sk < sz) {
            if (nums[sk] != 0) {
                nums[p] = nums[sk];
                p++, sk++;
            }
            else {
                --bk, sk++;
            }
        }
        while (p < sz)
            nums[p++] = 0;
    }
};

int main() {

    Solution s;
    vector<int> nums{0,1,0,3,11};
    s.moveZeroes(nums);
    cout << reduce(next(nums.begin()), nums.end(), to_string(nums[0]), [](string a, int b) { return move(a) + ", " + to_string(b); }) << endl;
};

File: challenge/leet/easy/q268.py
'''
 * 268. Missing Number
 * https://leetcode.com/problems/missing-number/
'''
import unittest
from typing import List


class Solution:
    def missingNumber(self, nums: List[int]) -> int:
        v = 0
        for x, n in enumerate(nums):
            v ^= (x+1) ^ n
        return v


class MyTestCase(unittest.TestCase):
    def test_something(self):
        s = Solution()
        self.assertEqual(2, s.missingNumber([0,1,3]))
        self.assertEqual(0, s.missingNumber([2,1,3]))
        self.assertEqual(6, s.missingNumber([2,1,3,5,0,7,4]))


if __name__ == '__main__':
    unittest.main()
